src/object/srv_obj.c

```diff
@@ -5296,3 +5302,297 @@ ds_obj_key2anchor_handler(crt_rpc_t *rpc)
+	opi.opi_flags = ocpi->ocpi_flags & ~ORF_LEADER;
+
+	otpa.opc = opc_get(rpc->cr_opc);
+	otpa.opi = &opi;
+	otpa.mbs = ocpi->ocpi_odm.odm_mbs;
+	otpa.ver = &ocpa->versions[dmi->dmi_tgt_id];
+	otpa.data = rpc;
+
+	rc = obj_tgt_punch_one(&otpa);
+
+	DL_CDEBUG(rc == 0 || rc == -DER_INPROGRESS || rc == -DER_TX_RESTART, DB_IO, DLOG_ERR, rc,
+		  "Collective punch obj shard "DF_UOID" with "DF_DTI" on rank %u tgt %u",
+		  DP_UOID(opi.opi_oid), DP_DTI(&opi.opi_dti), oct->oct_rank, dmi->dmi_tgt_id);
+
+	return rc;
+}
+
+static int
+obj_coll_punch_local(crt_rpc_t *rpc, uint32_t *version)
+{
+	struct obj_coll_punch_in	*ocpi = crt_req_get(rpc);
+	struct obj_coll_target		*oct = &ocpi->ocpi_tgts.ca_arrays[0];
+	struct obj_coll_punch_args	 ocpa = { 0 };
+	struct dss_coll_ops		 coll_ops = { 0 };
+	struct dss_coll_args		 coll_args = { 0 };
+	int				 len = oct->oct_bitmap_sz << 3;
+	int				 rc = 0;
+	int				 i;
+
+	ocpa.rpc = rpc;
+	D_ALLOC_ARRAY(ocpa.versions, len);
+	if (ocpa.versions == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	coll_ops.co_func = obj_coll_punch_one;
+	coll_args.ca_func_args = &ocpa;
+	coll_args.ca_tgt_bitmap = oct->oct_bitmap;
+	coll_args.ca_tgt_sz = oct->oct_bitmap_sz;
+
+	rc = dss_thread_collective_reduce(&coll_ops, &coll_args, DSS_ULT_DEEP_STACK);
+
+out:
+	if (ocpa.versions != NULL) {
+		for (i = 0, *version = 0; i < len; i++) {
+			if (*version < ocpa.versions[i])
+				*version = ocpa.versions[i];
+		}
+	}
+
+	DL_CDEBUG(rc == 0 || rc == -DER_INPROGRESS || rc == -DER_TX_RESTART, DB_IO, DLOG_ERR, rc,
+		  "Collective punch obj "DF_UOID" with "DF_DTI" on rank %u, got version %u",
+		  DP_UOID(ocpi->ocpi_oid), DP_DTI(&ocpi->ocpi_odm.odm_xid), oct->oct_rank,
+		  *version);
+
+	D_FREE(ocpa.versions);
+	return rc;
+}
+
+struct obj_coll_local_ult_args {
+	crt_rpc_t	*oclua_rpc;
+	ABT_future	 oclua_future;
+	uint32_t	 oclua_version;
+	int		 oclua_ret;
+};
+
+static void
+obj_coll_punch_ult(void *args)
+{
+	struct obj_coll_local_ult_args	*oclua = args;
+	int				 rc;
+
+	oclua->oclua_ret = obj_coll_punch_local(oclua->oclua_rpc, &oclua->oclua_version);
+
+	rc = ABT_future_set(oclua->oclua_future, NULL);
+	D_ASSERTF(rc == ABT_SUCCESS, "ABT_future_set failed: %d\n", rc);
+}
+
+static int
+obj_coll_punch_disp(struct dtx_leader_handle *dlh, void *arg, int idx, dtx_sub_comp_cb_t comp_cb)
+{
+	struct ds_obj_exec_arg		*exec_arg = arg;
+	crt_rpc_t			*rpc = exec_arg->rpc;
+	struct obj_coll_punch_in	*ocpi = crt_req_get(rpc);
+	struct obj_coll_target		*oct = &ocpi->ocpi_tgts.ca_arrays[0];
+	struct obj_punch_in		 opi = { 0 };
+	struct obj_coll_local_ult_args	 ult_arg = { 0 };
+	int				 rc = 0;
+	int				 rc1;
+
+	if (idx != -1)
+		return ds_obj_coll_punch_remote(dlh, arg, idx, comp_cb);
+
+	ult_arg.oclua_rpc = rpc;
+	rc = ABT_future_create(1, NULL, &ult_arg.oclua_future);
+	if (rc != 0)
+		goto out;
+
+	rc = dss_ult_create(obj_coll_punch_ult, &ult_arg, DSS_XS_IOFW,
+			    dss_get_module_info()->dmi_tgt_id, DSS_DEEP_STACK_SZ, NULL);
```

Comment:
Is this 'obj_coll_punch_ult" for local punch? Why we need to create an ULT on IO forward xstream to perform it?

Reply:
It is for the collective punch on other non-leader vos targets on the same local engine. That can be driven on helper XS/ULT.